new "type system concepts" section (#1743)
python/typing: docs/spec/generics.rst

  longer({1}, {1, 2})  # ok, return type set[int]
  longer([1], {1, 2})  # ok, return type a supertype of list[int] and set[int]

An upper bound cannot be combined with type constraints (as used in ``AnyStr``,

> see the example earlier); type constraints cause the inferred type to be

Reviewer: Off-topic (this should be tackled separately), but we shouldn't use `AnyStr` as an example in this document. It's deprecated on py313+ because it has a truly terrible name (it neither means "any string", nor has anything to do with `Any`), and scheduled for removal in 3.18. See https://github.com/python/cpython/issues/105578

Bound Rules
^^^^^^^^^^^

> ``T1``'s bound must be a :term:`consistent subtype` of ``T2``'s bound.

Author: Yeah, I think as I was editing I thought it read better, but on second look I don't see any good reason. I'll change it.

"""""""""""""""""""""""""""""""

If both ``bound`` and ``default`` are passed, ``default`` must be a

> :term:`consistent subtype` of ``bound``. If not, the type checker should

Reviewer 2: Why not "assignable to"?
```suggestion
:term:`assignable` to ``bound``. If not, the type checker should
```
(Also need to remove "a" on the previous line.)

^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Another use for ``Self`` is to annotate attributes. One example is where we

> have a ``LinkedList`` whose elements must be consistent subtypes of the current
> class.

Reviewer: ```suggestion
have a ``LinkedList`` whose elements must be
:term:`consistent subtypes <consistent subtype>` of the current class.
```

        def ordinal_value(self) -> str:
            return as_ordinal(self.value)

>     # Should not be OK because LinkedList[int] is not a consistent subtype of

Reviewer 2: ```suggestion
    # Should not be OK because LinkedList[int] is not assignable to
```

3. Create two specialized versions of the class. We'll refer to these as
``upper`` and ``lower`` specializations. In both of these specializations,
replace all type parameters other than the one being inferred by a dummy type
instance (a concrete anonymous class that is assumed to meet the bounds or
constraints of the type parameter). In the ``upper`` specialized class,

> specialize the target type parameter with an ``object`` instance. This

Author: I agree that I find the usage of "instance" here not super clear, but I'd rather leave that for a separate pull request, since it's not directly related to this PR.
